ubi_manifest/worker/tasks/content_audit.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import os
+
+from pubtools.pulplib import Criteria
+
+from ubi_manifest.worker.tasks.celery import app
+from ubi_manifest.worker.tasks.depsolve import filter_whitelist, get_content_config
+from ubi_manifest.worker.tasks.depsolver.models import UbiUnit
+from ubi_manifest.worker.tasks.depsolver.pulp_queries import (
+    search_modulemd_defaults,
+    search_modulemds,
+    search_rpms,
+)
+from ubi_manifest.worker.tasks.depsolver.ubi_config import UbiConfigLoader
+from ubi_manifest.worker.tasks.depsolver.utils import (
+    create_or_criteria,
+    get_criteria_for_modules,
+    is_blacklisted,
+    keep_n_latest_modulemd_defaults,
+    keep_n_latest_modules,
+    keep_n_latest_rpms,
+    make_pulp_client,
+    parse_blacklist_config,
+)
+
+_LOG = logging.getLogger(__name__)
+MAX_WORKERS = int(os.getenv("UBI_MANIFEST_CONTENT_AUDIT_WORKERS", "8"))
+
+
+@app.task  # type: ignore [misc]  # ignore untyped decorator
+def content_audit_task() -> None:
+    """
+    This task checks that all available content is up-to-date, that whitelisted
+    content is present, and that blacklisted content is absent.
+    """
+
+    config_loaders = [UbiConfigLoader(url) for url in app.conf.content_config.values()]
+
+    with make_pulp_client(app.conf) as client:
+        for out_repo in client.search_repository(
+            Criteria.with_field("ubi_population", True)
+        ):
+            # get all relevant units currently on output repo
+            out_rpms = search_rpms([Criteria.true()], [out_repo]).result()
+            out_mds = search_modulemds([Criteria.true()], [out_repo]).result()
+            out_mdds = search_modulemd_defaults([Criteria.true()], [out_repo]).result()
+
+            seen_units = set()
+            output_whitelist = set()
+
+            for in_repo in client.search_repository(
+                Criteria.with_id(out_repo.population_sources)
+            ):
+                # get all corresponding units currently on input repo
+                in_rpms_ft = search_rpms(
+                    create_or_criteria(["name"], [(rpm.name,) for rpm in out_rpms]),
+                    [in_repo],
+                )
+                in_mds_ft = search_modulemds(
+                    get_criteria_for_modules(list(out_mds)), [in_repo]
+                )
+                in_mdds_ft = search_modulemd_defaults(
+                    get_criteria_for_modules(list(out_mdds)), [in_repo]
+                )
+
+                # check that all content is up-to-date
+                in_rpms = list(in_rpms_ft.result())
+                keep_n_latest_rpms(in_rpms)
+                for in_rpm in in_rpms:
```

Comment:
Although this whole task won't be critical for performance, these comparisons are very inefficient.
Have you tried running this on production-like data?

Reply:
I've made changes to strike seen units from the output list so they won't be checked again on the next input unit iteration.

As for local testing, I assume I'd have to run the container to use this task. That's manageable but I'm not sure where to find real content config maps. For the URLs, is your example repo okay to use (rbikar/ubi-config-examples)?

Reply:
@rbikar, I can't seem to get the pod running locally. It's built and brought up but it doesn't give me a working hostname (none of the usual addresses).  I'm not sure how I'd check the task with a working pod anyway. The task doesn't have an endpoint so it can't be called manually, would just need to let it run for a few hours I suppose. But then I'm not sure if there would be visible signs the task has run. You could monitor it but if nothing is outdated, missing, etc. then no logs would be written.